crates/sdl-encoder/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+    }
+}
+
+impl Display for SchemaDef {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(description) = &self.description {
+            writeln!(f, "\"\"\"\n{}\n\"\"\"", description)?;
+        }
+
+        let mut fields = String::new();
+        for field in &self.fields {
+            fields += &format!("\n{}", field.to_string());
+        }
+
+        write!(f, "schema {{")?;
+        write!(f, "{}", fields)?;
+        writeln!(f, "\n}}")
+    }
+}
+
+/// Define whether a Field is a vector.
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub enum VecState {
+    /// The type is not a vector: e.g. `String`.
+    None,
+    /// The type is a nullable vector: e.g. `[String]`.
+    Nullable,
+    /// The type is a non nullable vector: e.g. `[String]!`.
+    NonNullable,
+}
+/// Field in a given SDL type.
+#[derive(Debug, PartialEq, Clone)]
+pub struct Field {
+    description: Option<String>,
+    name: String,
+    type_: String,
+    default: Option<String>,
+    is_nullable: bool,
+    vec_state: VecState,
+}
+
+impl Field {
+    /// Create a new instance of Field.
+    pub fn new(name: String, type_: String) -> Self {
+        Self {
+            description: None,
+            name,
+            type_,
+            default: None,
+            is_nullable: true,
+            vec_state: VecState::None,
+        }
+    }
+
+    /// Set the field's description.
+    pub fn description(&mut self, description: String) {
+        self.description = Some(description);
+    }
+
+    /// Set the field's default.
+    pub fn default(&mut self, default: String) {
+        self.default = Some(default);
+    }
+
+    /// Set the field's is nullable.
+    pub fn is_nullable(&mut self, nullable: bool) {
+        self.is_nullable = nullable;
+    }
+
+    /// Set the field's vec state.
+    pub fn vec_state(&mut self, vec_state: VecState) {
+        self.vec_state = vec_state;
+    }
+}
+
+impl Display for Field {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(description) = &self.description {
+            // Let's indent description on a field level for now, as all fields
+            // are always on the same level and are indented by 2 spaces.
+            writeln!(f, "  \"\"\"\n  {}\n  \"\"\"", description)?;
+        }
+        let null = if self.is_nullable { "!" } else { "" };
+        let type_ = match self.vec_state {
+            VecState::None => format!("{}{}", &self.type_, null),
+            VecState::Nullable => format!("[{}{}]", &self.type_, null),
+            VecState::NonNullable => format!("[{}{}]!", &self.type_, null),
+        };
+        let default = if let Some(default) = &self.default {
+            match self.vec_state {
+                VecState::None => format!(" = {}{}", default, null),
+                VecState::Nullable => format!(" = [{}{}]", default, null),
+                VecState::NonNullable => format!(" = [{}{}]!", default, null),
+            }
```

Comment:
I am currently encoding `default` values based on the shape of the original type, so let's say the field looks like:
```js
key: ValueEnum!
```
The default value will look like:
```js
key: ValueEnum! = EnumVersionOne!
```
If the field looks like this:
```js
key: [ValueEnum!]!
```
The default would look like this:
```js
key: [ValueEnum!]! = [EnumVersionOne!]!
```

Does this seem correct? Should I be making any adjustments?

Reply:
I hope I understand the question correctly, but fields by themselves don't have default values. Only field arguments do (and variables, but that isn't relevant for schemas).

As for how these default values are represented, you may not want to use a `string` for this but rely on a more structured way of representing values. In Apollo iOS we use [an enum](https://github.com/apollographql/apollo-ios/blob/main/Sources/ApolloCodegenLib/Frontend/GraphQLValue.swift) for this, which is similar to [what `graphql-js` does in its AST](https://github.com/graphql/graphql-js/blob/735975d4904c5d9cc93d0c0a02af181e8f02dfba/src/language/ast.d.ts#L309-L318).

Reply:
I am sorry, I did mean `value`, or field argument in a given field. I was looking at a few SDLs and wanted to find out what variations there are for a way to display a default. I found things like this:

```js
input DirectorySortInput {
  fields: [DirectoryFieldsEnum]
  order: [SortOrderEnum] = [ASC] // this is a default assignment, correct?
}
```
and
```js
type ImageSharp implements Node {
    srcSetBreakpoints: [Int] = [] // so are these three?
    toFormat: ImageFormat = NO_CHANGE
    toFormatBase64: ImageFormat = NO_CHANGE
}
```

And I am wondering what other ways there could be to set a default, in case my assumption is incorrect.

As for default value representation, everything is going to be a `String` as this is a serialiser and all it concerns itself with is how to stringify the data.

Reply:
The second example isn't actually valid, because fields on object types and interfaces can't have default values. Only fields on input objects or field arguments can.

So this is an additional place where default values might come in:
```
type Product {
  weight(unit: WeightUnit = KILOS): Float
}
```

Reply:
For that reason, `graphql-js` differentiates between `GraphQLField` (which doesn't have a default value) and `GraphQLInputField` (which does).

Reply:
Ah ok! gotcha. I have not seen a default for a field argument example yet, so this is useful.

Reply:
Given the example you used what are the different components in the statement:
```js
type Product {
  weight(unit: WeightUnit = KILOS): Float
}
```
i.e. what would you call  these components: `unit`, `WeightUnit` and `KILOS`?